### PR TITLE
feat(external-router): process redirects and errors

### PR DESCRIPTION
**Pull request by damienwebdev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently, redirects cause an infinite loop.

Fixes: #1499 

## What is the new behavior?
We no longer get stuck in an infinite loop.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: libs/external-router/driver/magento/src/processors/process-redirect.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  DaffExternalRouterTemporaryRedirectError,
+} from '@daffodil/external-router/driver';
+
+import { MagentoUrlResolverResponse } from '../model/public_api';
+
+/**
+ * Processes Magento responses and throws the appropriate error so that External Router
+ * can handle redirects.
+ */
+export const processRedirectResponse = (response: ApolloQueryResult<MagentoUrlResolverResponse>) => {
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/external-router/driver/magento/src/processors/process-redirect.ts:14]: I agree but I prefer `magentoProcessRedirectResponse` to stay consistent

**Comment by damienwebdev** [comment]
@lderrickable @griest024 I've reworked this fairly significantly, please take another look. I like this version much better. It allows the drivers to act like regular HttpResponse results and defers managing those results to the guard.

File: libs/external-router/routing/src/guard/existence.guard.ts [hunk 2]
```diff
@@ -54,12 +63,40 @@ export class DaffExternalRouterExistenceGuard implements CanActivate {
...
+      // Handle redirects
       catchError(error => {
+        if(
+          !(error instanceof DaffExternalRouterPermanentRedirectError) &&
+          !(error instanceof DaffExternalRouterTemporaryRedirectError)) {
+          throw error;
+        }
+
+        //On platform browser, we need to add a redirect for the original url and redirect to the new url.
+        this.externalRouter.addRedirect(state.url, error.redirectUrl);
```
Comment (damienwebdev) [inline libs/external-router/routing/src/guard/existence.guard.ts:85]: `state.url` may not be safe to insert here as it contains queries and hashes.

File: libs/external-router/routing/src/guard/existence.guard.ts [hunk 3]
```diff
@@ -54,12 +63,40 @@ export class DaffExternalRouterExistenceGuard implements CanActivate {
...
+        return of(this.router.parseUrl(error.redirectUrl));
+      }),
+      // Handle "404"
+      catchError((error) => {
+        if(!(error instanceof DaffExternalRouterNotFoundError)) {
+          throw error;
+        }
+
+        this.externalRouter.addRedirect(state.url, this.config.failedResolutionPath);
```
Comment (damienwebdev) [inline libs/external-router/routing/src/guard/existence.guard.ts:94]: `state.url` may not be safe to insert here as it contains queries and hashes.

File: libs/external-router/routing/src/guard/existence.guard.ts (hunk 2 above)
Comment (griest024) [reply libs/external-router/routing/src/guard/existence.guard.ts:85]: yeah because that means that the path will only match with the query strings, right?

**Review by griest024 — CHANGES_REQUESTED** [review]

File: libs/external-router/routing/src/errors/client-error.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import {
+  DaffError,
+  DaffInheritableError,
+} from '@daffodil/core';
+
+/**
+ * An error thrown when the driver needs a URI to be redirected permanently
+ * to another location.
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/external-router/routing/src/errors/client-error.ts]: Fixed in abe8b76cc3adb9b8ac0ef4e98dff25987494542e

File: libs/external-router/routing/src/guard/existence.guard.ts (hunk 2 above)
Comment (damienwebdev) [reply libs/external-router/routing/src/guard/existence.guard.ts]: Fixed in abe8b76cc3adb9b8ac0ef4e98dff25987494542e -- @griest024 can you give it a once over to verify.

File: libs/external-router/driver/src/public_api.ts [hunk 5]
```diff
@@ -6,3 +6,4 @@ export {
   DaffExternalRouterDriverInterface,
   DaffExternalRouterDriver,
 } from './interfaces/external-router-driver.interface';
+export { NOT_FOUND_RESOLUTION } from './not-found-resolution';
```
Comment (griest024) [inline libs/external-router/driver/src/public_api.ts:9]: this should have a daff or magento prefix since its exported

**Review by griest024 — CHANGES_REQUESTED** [review]

File: libs/external-router/driver/src/public_api.ts (hunk 5 above)
Comment (damienwebdev) [reply libs/external-router/driver/src/public_api.ts]: 28c849b8

**Review by griest024 — APPROVED** [review]

**Pull request merged**